spinalcordtoolbox/qmri/mt.py

```diff
@@ -44,8 +44,14 @@ def compute_mtr(nii_mt1, nii_mt0, threshold_mtr=100):
     :param threshold_mtr: float: value above which number will be clipped
     :return: nii_mtr
     """
+    # Convert input to avoid numerical errors from int16 data
+    # Related issue: https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3636
+    nii_mt1.change_type('float32')
+    nii_mt0.change_type('float32')
```

Comment:
Just to be absolutely thorough, I copied the 2 new tests over to `master` (to make sure that they properly capture the initial buggy behavior reported in #3636). 

However, the results of the tests are the same on both `master` and this branch. In other words, the `float32` changes seem to have no impact, at least for the test data we've tried.

In hindsight, I realize that we probably should have been working with this user's exact data to try and isolate the problem precisely. @jcohenadad, is that something you would be able to share privately?

Reply:
No problem whatsoever! This is partially on me, too -- I wrote the initial test with the `sct_testing_data` files as well, without considering the problem at hand. :sweat_smile: 

Regardless, the time was not wasted at all! What you've learned will still be very valuable for working with image data in the future. :)

Reply:
I can confirm that the `float32` change does indeed have a large effect on the user's data. Note the screenshots below:

<details>

sct_compute_sct on `master`:

![master](https://user-images.githubusercontent.com/16181459/151027144-b60ef3c0-c92d-4004-95a2-661c04fed9fb.png)

sct_compute_sct with the float32 fix:

![with_float32_fix](https://user-images.githubusercontent.com/16181459/151027150-88284798-538e-4919-aa89-6976592c4ab0.png)

Difference between the two images:

![diff](https://user-images.githubusercontent.com/16181459/151028060-c8fdef71-f4b1-40df-9d54-ed06e6ddc605.png)

</details>

Investigating deeper, the root cause of the issue here is actually... dead simple. :facepalm: 

It's not _just_ that the user's data is `int16`, but that the user's data spans the range `[0, 4095]`:

<details>

![Screenshot from 2022-01-25 12-36-15](https://user-images.githubusercontent.com/16181459/151029262-0b604e45-6d7f-4201-a46b-2c3030282cb5.png)

</details>

When we combine these large values with the [`100 *` step](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/26b7fe7aa1784f61de4058b5d14a29789f9defe0/spinalcordtoolbox/qmri/mt.py#L50) in our MTR calculation code, it results in the data getting clipped by the `2^15` limits of the `int16` datatype:

<details>

![Screenshot from 2022-01-25 12-39-13](https://user-images.githubusercontent.com/16181459/151029745-cf070329-7cd6-4543-84ac-9fd107e9c6e4.png)

</details>

By changing the datatype to `float32`, the data  is no longer clipped, and we can compute the MTR as expected:

<details>

![Screenshot from 2022-01-25 12-40-44](https://user-images.githubusercontent.com/16181459/151030037-4cf4583c-116b-4e76-89f9-7cb3f92a564d.png)

</details>

So, the original fix works as intended! And, the only reason our tests didn't capture the original buggy behavior is that our values weren't large enough. :laughing:

I'll fix up the tests so that they fail on `master` but pass on this branch, and we should be good to go. :)

Reply:
@joshuacwnewton Good catch! In this context should we enhance a bit my fix and transform ALL the images input into a biggest dtype before calculation ? Today it was a bug due to int16 but tomorrow it could be an error with a int64 for example.

```
(2, 'uint8', np.uint8, "NIFTI_TYPE_UINT8"),\
(4, 'int16', np.int16, "NIFTI_TYPE_INT16"),\
(8, 'int32', np.int32, "NIFTI_TYPE_INT32"),\
(16, 'float32', np.float32, "NIFTI_TYPE_FLOAT32"),\
(32, 'complex64', np.complex64, "NIFTI_TYPE_COMPLEX64"),\
(64, 'float64', np.float64, "NIFTI_TYPE_FLOAT64"),\
(256, 'int8', np.int8, "NIFTI_TYPE_INT8"),\
(512, 'uint16', np.uint16, "NIFTI_TYPE_UINT16"),\
(768, 'uint32', np.uint32, "NIFTI_TYPE_UINT32"),\
(1024,'int64', np.int64, "NIFTI_TYPE_INT64"),\
(1280, 'uint64', np.uint64, "NIFTI_TYPE_UINT64"),\
(1536, 'float128', _float128t, "NIFTI_TYPE_FLOAT128"),\
(1792, 'complex128', np.complex128, "NIFTI_TYPE_COMPLEX128"),\
(2048, 'complex256', _complex256t, "NIFTI_TYPE_COMPLEX256"),
```

Reply:
That's a good question... Perhaps a bit out of scope for this PR, though? I feel like I want to give it more time and thought. :thinking:

Reply:
The actual fix is an enhancement, and it will fix the problem for the user that created the issue. 
We can take more time and create a new issue for the generic problem.

Reply:
As an aside, I've fixed up the test to use high-valued data in ab3d0be!

And, just to demonstrate that this modified test does indeed capture the buggy behavior on `master`, here's what the results look like when the data gets clipped:

![Screenshot from 2022-01-25 13-47-55](https://user-images.githubusercontent.com/16181459/151039701-f43a87ba-1155-4354-bfc2-55888ab08cc7.png)

On this branch, however, the two arrays are identical. :smile: